### ISSUE TITLE
Error: Cannot find module next/link imported from @flowershow/core/index.js. Did you mean to import next/link.js?

### ISSUE THREAD

**Issue by olayway** [flowershow/flowershow#369]
After creating `@flowershow/core` package and using it as a dependency of the `@flowershow/template` package, the following error shows up when running E2E tests or creating new Flowershow sites with our CLI:

```
Uncaught Error: Cannot find module '/home/ola/Projects/lifeitself/my-site/.flowershow/node_modules/next/link' imported from /home/ola/Projects/lifeitself/my-site/.flowershow/node_modules/@flowershow/core/index.js
Did you mean to import next/link.js?
    at new NodeError (node:internal/errors:387:5)
    at finalizeResolution (node:internal/modules/esm/resolve:330:11)
    at moduleResolve (node:internal/modules/esm/resolve:907:10)
    at defaultResolve (node:internal/modules/esm/resolve:1115:11)
    at nextResolve (node:internal/modules/esm/loader:163:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:837:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
    at link (node:internal/modules/esm/module_job:75:36)
```

The error doesn't occur when serving the template with `pnpm nx serve template` though.

**Comment by olayway** [comment]
@khalilcodes here is the reference page of the rollup executor we're using in the core package: https://nx.dev/packages/web/executors/rollup

Atm it's using a default rollup config returned by this function:
https://github.com/nrwl/nx/blob/master/packages/react/plugins/bundle-rollup.ts

You can also create your own config or extend the default one. This article may help you if you want to do this: https://antonball.medium.com/extending-rollup-configuration-for-nx-a568cec16bf4

**Comment by olayway** [comment]
Last thing I tried was generating both cjm and esm (https://github.com/flowershow/flowershow/commit/1db5cbbb5f66c6acddfa387861eb575f7b636850) but it didn't have any effect.

**Comment by rufuspollock** [comment]
Where is next/link getting imported? Is there a temporary workaround eg importing next/link in template?

First step here would be to remove things until this starts working. Then rebuild from a working state. 

Btw these issues would show up with eg Pliny so perhaps we could examine how they build to have this work.

**Comment by rufuspollock** [comment]
Where is next/link getting imported? Is there a temporary workaround eg importing next/link in template?

First step here would be to remove things until this starts working. Then rebuild from a working state. 

Btw these issues would show up with eg Pliny so perhaps we could examine how they build to have this work.

**Comment by olayway** [comment]
I have submitted a question on Stack Overflow: https://stackoverflow.com/questions/74900445/error-cannot-find-module-node-modules-next-link-imported-from-did-you
And I've also asked a question on Next.js's discord channel: https://discord.com/channels/752553802359505017/1055551199807549440/1055551199807549440

**Comment by olayway** [comment]
Until this is solved I'll revert the version on main to the working version before the monorepo refactor and I'll move what we have on `main` atm to `dev` branch.

**Comment by aabounegm** [comment]
I faced this issue today trying to run `npx flowershow@latest export` (or `build`). Was the rollback pushed to npm or should this issue not be there to begin with?

_Environment_:
| flowershow | OS | Node | npm |
| ---------- | -- | ---- | --- |
| latest (`@0.1.2`) | Windows 10 22H2 | `v18.12.0` | `8.19.2` |

**Comment by olayway** [comment]
@aabounegm the rollback has been pushed to our main branch (this is where the CLI get's the Flowershow template from) ~ a week ago. Please let me know if it works for you now.

**Comment by olayway** [comment]
@rufuspollock 
I've managed to get past that error by explicitly specifying file extensions in any imports from `next/link`, `next/router`, and `next/head` **in the Flowershow core package.**

I'm not sure why (I may lack some basic knowledge about module resolution here) but it seems Next can only handle imports without file extensions for import statements used directly in the Next.js project, e.g. in our template, but not when such import statements are used in a dependency of this project.

Originally, with import statements without `.js` extensions, this was the resulting `index.js` file of the bundled core package:

```js
...
import 'react/jsx-runtime';
import 'react';
import 'next/link'; <-- no extension; caused the above error
import 'next-themes';
import 'next/dynamic';  <-- no extension; caused the above error
import '@headlessui/react';
import 'next/router';  <-- no extension; caused the above error
import 'next/head';  <-- no extension; caused the above error
import 'mdx-mermaid/lib/Mermaid';
import 'clsx';
```

After adding file extensions to any imports from `next/XXX` files, the above file looks like this:

```js
...
import 'react/jsx-runtime';
import 'react';
import 'next/link.js'; <-- no error now
import 'next-themes';
import 'next/dynamic.js';  <-- no error now
import '@headlessui/react';
import 'next/router.js';  <-- no error now
import 'next/head.js';  <-- <-- no error now
import 'mdx-mermaid/lib/Mermaid';
import 'clsx';
```

What's interesting is that only imports from `next` package caused these errors. Why? Because all the other packages have `exports` field specified in their `package.json` files, e.g. in `react`'s `package.json`:

```
  "exports": {
    ".": {
    ...
      "default": "./index.js"
    },
    ...
    "./jsx-runtime": "./jsx-runtime.js",
    ...
  },

```
... which enables resolving both imports from `react` (the default export filed) and from `react/jsx-runtime`.

**Comment by olayway** [comment]
@rufuspollock this is solved, but now I'm working on this error, but this should be much easier to solve :)

![image](https://user-images.githubusercontent.com/52197250/211323580-3e27ffd6-6435-44a6-b4c9-ea7570517fa7.png)

**Comment by olayway** [comment]
Ok i think this is solved as well.

**Comment by aabounegm** [comment]
@olayway Thanks for the response, I indeed noticed the rollback happened after I wrote my comment above. The build is now failing for a different reason, but I think it deserves a separate issue.

Also, you are absolutely right about the module resolution strategy. ESM necessitates the use of file extension (Check [Node.js ESM docs](https://nodejs.org/api/esm.html#mandatory-file-extensions)), except for packages that define the `exports` field. In other words, adding `.js` is not just a workaround, but the right solution.

> "we can only import files from next package without using file extensions for import statements used directly in the Next.js project, ..., but not when such import statements are used in a dependency of this project"

I'm not totally sure about the specifics in this case (haven't used Next.js), but it depends on the bundler configuration. Most likely, Next configures Webpack (the default bundler when using the `next` CLI) to handle CommonJS imports by default, but you would need to explicitly add CommonJS support to your Rollup configuration (using [`@rollup/plugin-commonjs`](https://www.npmjs.com/package/@rollup/plugin-commonjs)) when using Next.js as a dependency since it is shipped as a CommonJS module.
